### PR TITLE
stable, dev PyTorch in Dockerfile and conda gh actions

### LINKED ISSUE #2090: [test] Add tests against PT nightly using GPUs
Testing against nightly build.

### PR DESCRIPTION
**Pull request by ydcjeff**
## What does this PR do?

Add Docker image installed with pytorch nightly, PyTorch & Conda gh action tested with pytorch nightly

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: dockers/cuda-nightly/Dockerfile
Comment (ydcjeff) [inline dockers/cuda-nightly/Dockerfile:4]: Is this the correct image for installing various PyTorch versions?

**Comment by ydcjeff** [comment]
The new added github actions will be merged into `docker-builds.yml` if approved

File: dockers/cuda-nightly/Dockerfile
Comment (ydcjeff) [inline dockers/cuda-nightly/Dockerfile:51]: Do we need to install amp for pytorch >=1.6?

Comment (williamFalcon) [reply dockers/cuda-nightly/Dockerfile:51]: yes... because some 1.6 users still like apex

Comment (ydcjeff) [reply dockers/cuda-nightly/Dockerfile:51]: alright.

File: dockers/cuda-nightly/Dockerfile
Comment (williamFalcon) [reply dockers/cuda-nightly/Dockerfile:4]: @Borda

**Comment by Borda** [comment]
maybe but in Conda env we are running only CPU tests compare to the CUDA-pip image we run in Drone with the full suite case

**Comment by ydcjeff** [comment]
Found out pytorch CUDA image itself is also using conda, so go with it right now.
would you like to change to pip?

**Comment by Borda** [comment]
> Found out pytorch CUDA image itself is also using conda, so go with it right now.
> would you like to change to pip?

Well but is conda activated by default?

**Comment by ydcjeff** [comment]
> > Found out pytorch CUDA image itself is also using conda, so go with it right now.
> > would you like to change to pip?
> 
> Well but is conda activated by default?

yes

**Comment by ydcjeff** [comment]
for GPU testing with stable and dev PyTorch on Drone, shall we do testing in parallel or one after another?
I go for one after another
WDYT?

**Comment by ydcjeff** [comment]
@Borda I have noticed that `PyTorch & Conda` github actions is installing gpu version of pytorch while we are not testing with gpu. Is this for some reasons or shall i add `cpuonly` package in `environment.yml` to save a few pkg installing minutes?

**Comment by ydcjeff** [comment]
@Borda mind review this so far? 
so that i can proceed https://github.com/PyTorchLightning/pytorch-lightning/pull/3074#issuecomment-681859896 and https://github.com/PyTorchLightning/pytorch-lightning/pull/3074#issuecomment-680894091

**Comment by Borda** [comment]
btw, is it still wip?

**Comment by ydcjeff** [comment]
> btw, is it still wip?

I haven't worked on Drone CI for GPU testing, so can be said wip
waiting for review of current `base-cuda` Dokerfile

**Comment by ydcjeff** [comment]
@Borda do I need to change this docker image?
https://github.com/PyTorchLightning/pytorch-lightning/blob/64ee25ed9cd1d44bb276b61683bbdfee0b00ff9d/.drone.yml#L23

**Comment by Borda** [comment]
> @Borda do I need to change this docker image?
> https://github.com/PyTorchLightning/pytorch-lightning/blob/64ee25ed9cd1d44bb276b61683bbdfee0b00ff9d/.drone.yml#L23

you cannot until it is build for the first time and push to the docker hub...

**Comment by ydcjeff** [comment]
> > @Borda do I need to change this docker image?
> > https://github.com/PyTorchLightning/pytorch-lightning/blob/64ee25ed9cd1d44bb276b61683bbdfee0b00ff9d/.drone.yml#L23
> 
> you cannot until it is build for the first time and push to the docker hub...

Ok then, shall we build the docker from this PR without merging into master?
OR build the docker from master and make another follow PR of this?

**Comment by ydcjeff** [comment]
@Borda ready to review

**Comment by Borda** [comment]
@ydcjeff I am sorry for the delay, hope to check it over the weekend...

**Comment by Borda** [comment]
let's wait this build is merged #3383

File: dockers/base-cuda/Dockerfile
Comment (Borda) [inline dockers/base-cuda/Dockerfile:16]: let's add defaults

File: .github/workflows/docker-builds.yml
Comment (Borda) [inline .github/workflows/docker-builds.yml:98]: 👍

File: .github/workflows/pt-conda.yml
Comment (Borda) [inline .github/workflows/pt-conda.yml:35]: `pytorch` is also a package name, so we hope that the channels ale always before packages...

File: .github/workflows/pt-conda.yml
Comment (Borda) [inline .github/workflows/pt-conda.yml:33]:
```suggestion
      if: matrix.pytorch-version >= 1.7
```
to be a bit more generic

File: README.md
Comment (Borda) [inline README.md:73]: all other lines has to have this extra column

File: environment.yml
Comment (Borda) [inline environment.yml:27]: why is this needed?

**Review by Borda — CHANGES_REQUESTED** [review]

File: .github/workflows/pt-conda.yml
Comment (ydcjeff) [reply .github/workflows/pt-conda.yml:35]: Yes, I am only replacing `pytorch` channel with `pytorch-nightly` channel

File: dockers/base-cuda/Dockerfile
Comment (ydcjeff) [reply dockers/base-cuda/Dockerfile:16]: so defaults are pytorch 1.6 and python 3.7?

File: README.md
Comment (ydcjeff) [reply README.md:73]: yep, will update it

File: environment.yml
Comment (ydcjeff) [reply environment.yml:27]: `PyTorch & Conda` CI is installing gpu version of pytorch whereas we are only testing in cpu mode, right?
so, adding that (installing cpu version of pytorch only) will save a few minutes of downloading and installing pkgs

Comment (Borda) [reply environment.yml:27]: but the env shall be generic not just for local CPU testing and we assume that most users have GPU, so if this is needed let's add it in the CI not here

File: .github/workflows/pt-conda.yml
Comment (Borda) [reply .github/workflows/pt-conda.yml:35]: it is fine, just saying let's be careful 🐰

File: dockers/base-cuda/Dockerfile
Comment (Borda) [reply dockers/base-cuda/Dockerfile:16]: I would say so, but you can put anything, it serves as an example for a user what he can add and what is the format...

File: environment.yml
Comment (ydcjeff) [reply environment.yml:27]: Okay I will apply the suggestions once #3383 is merged

File: .github/workflows/pt-conda.yml
Comment (ydcjeff) [reply .github/workflows/pt-conda.yml:35]: Okay

**Comment by Borda** [comment]
@ydcjeff blocking #3383 is merged, can you pls rebase so we can finish this one...

**Comment by ydcjeff** [comment]
@Borda good to go

File: .github/workflows/ci_pt-conda.yml
Comment (Borda) [inline .github/workflows/ci_pt-conda.yml:33]:
```suggestion
        # NOTE: this requires thet the channel is prensted in the yaml berore packages
        python -c "fname = 'environment.yml' ; req = open(fname).read().replace('pytorch', 'pytorch-nightly', 1) ; open(fname, 'w').write(req)"
```

File: .github/workflows/ci_pt-conda.yml
Comment (Borda) [inline .github/workflows/ci_pt-conda.yml:35]:
```suggestion
    - name: Setup PyTorch version
```

**Review by Borda — COMMENTED** [review]
just thinking that the image is very similar to the conda one, see would it make sense to later build the `Conda image on base CUDA?

**Comment by ydcjeff** [comment]
> just thinking that the image is very similar to the conda one, see would it make sense to later build the `Conda image on base CUDA?

yes, that's a great option.

**Review by Borda — APPROVED** [review]

**Comment by ydcjeff** [comment]
@Borda Don't we need to install `extra.txt` and amp?

**Comment by Borda** [comment]
@tgaddair mind help with building Horovod for Conda env?

**Comment by Borda** [comment]
> @Borda Don't we need to install amp?

in Conda we have it already as package
https://github.com/PyTorchLightning/pytorch-lightning/blob/93cf6d00541c1e6c6aef30010e62ed471c423af5/environment.yml#L35

**Comment by tgaddair** [comment]
Hey @Borda, what's the issue you're seeing with Horovod?

**Comment by Borda** [comment]
It was failing without special reason...

**Comment by tgaddair** [comment]
> It was failing without special reason...

Is it still failing?  Looks like tests are passing, did you disable something?

**Comment by ydcjeff** [comment]
@tgaddair Thank you for your reply.

At first, installing horovod was failing with if we set ``MAKEFLAG`` resulting in the error of unknow argument of VERBOSE=1

Accidently, I found out your PR on horovod and temporarily commnet out ``MAKEFLAGS`` and it worked.

Is this the correct way to fix for now?
Tests are installing horovod 0.20.

**Comment by tgaddair** [comment]
Hey @ydcjeff, glad you found that fix.  We will probably release v0.20.1 in a week or two, so I will try and remember to update this when we do.

Yes, unsetting `MAKEFLAGS` is a workaround.  Another option is to install with `pip install --global-option="--quiet" `, so the verbose flag does not get set.

**Review by Borda — APPROVED** [review]
LGTM, just the build seems to be failing on PT 1.7

**Comment by Borda** [comment]
> Hey @ydcjeff, glad you found that fix. We will probably release v0.20.1 in a week or two, so I will try and remember to update this when we do.
> 
> Yes, unsetting `MAKEFLAGS` is a workaround. Another option is to install with `pip install --global-option="--quiet" `, so the verbose flag does not get set.

wound this global work also when it is installed with Conda?

**Comment by ydcjeff** [comment]
> LGTM, just the build seems to be failing on PT 1.7

that is b/c `torchvision` was installed as 0.2.1, haven't faced that kind of lower version in previous commits.

**Comment by ydcjeff** [comment]
> > Hey @ydcjeff, glad you found that fix. We will probably release v0.20.1 in a week or two, so I will try and remember to update this when we do.
> > Yes, unsetting `MAKEFLAGS` is a workaround. Another option is to install with `pip install --global-option="--quiet" `, so the verbose flag does not get set.
> 
> wound this global work also when it is installed with Conda?

I think it works

**Comment by Borda** [comment]
@ydcjeff seems that tests for pt 1.7 are failing so I would recommend to keep the CI ready and just comment for now...
then make another PR which would fix the failing test and uncomment the testing

**Review by teddykoker — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]
LGTM, great job

**Review by rohitgr7 — APPROVED** [review]
All good. Nice work!

**Pull request merged**
